utils/buildcli/buildcli/clang.py

```diff
@@ -0,0 +1,85 @@
+from __future__ import absolute_import
+
+from collections import namedtuple
+import platform
+import subprocess
+
+from . import xcrun
+from .which import which
+
+# host_clang() prefers versions of Clang in this order.
+# Note that this array must be updated when new, major versions of Clang
+# are released.
+_CLANG_SUFFIXES = ['38', '37', '36', '35']
+
+
+class CompilerExecutable(namedtuple('CompilerExecutable', 'cc cxx')):
+    """
+    A named tuple consisting of two paths:
+
+    1. 'cc' is the path to a program used for compiling C.
+    2. 'cxx' is the path to a program used for compiling C++.
+    """
+    @property
+    def is_valid(self):
+        """
+        Return a boolean value indicating whether this instance represents
+        a valid C and C++ compiler pair.
+        """
+        return self.cc is not None and self.cxx is not None
+
+
+def _freebsd_release_date():
+    """
+    Return the release date for FreeBSD operating system on this host.
+    If the release date cannot be ascertained, return None.
+    """
+    try:
+        # For detail on `sysctl`, see:
+        # http://www.freebsd.org/cgi/man.cgi?sysctl(8)
+        return int(subprocess.check_output(
+            ['sysctl', '-n', 'kern.osreldate']).rstrip())
+    except subprocess.CalledProcessError:
+        return 0
+
+
+def _first_clang(suffixes=_CLANG_SUFFIXES):
+    """
+    Return a CompilerExecutable with the first available versions of clang
+    and clang++, searching in the order of the given suffixes.
+
+    If no Clang executables are found, return a CompilerExecutable with
+    None as values.
+    """
+    for suffix in suffixes:
+        cc_path = which('clang{}'.format(suffix))
+        cxx_path = which('clang{}'.format(suffix))
+        if cc_path and cxx_path:
+            return CompilerExecutable(cc=cc_path, cxx=cxx_path)
+
+    return CompilerExecutable(cc=None, cxx=None)
+
+
+def host_clang():
+    """
+    Return a CompilerExecutable for the host platform.
+    If no appropriate compilers can be found, return a CompilerExecutable
+    with None as values.
+    """
+    if platform.system() == 'Darwin':
+        return CompilerExecutable(cc=xcrun.find('clang'),
+                                  cxx=xcrun.find('clang++'))
+    elif platform.system() == 'FreeBSD':
+        # See: https://github.com/apple/swift/pull/169
+        # Building Swift from source requires a recent version of the Clang
+        # compiler with C++14 support.
+        freebsd_release_date = _freebsd_release_date()
+        if freebsd_release_date and freebsd_release_date >= 1100000:
+            # On newer releases of FreeBSD, the default Clang is sufficient.
+            return CompilerExecutable(cc='clang', cxx='clang++')
+        else:
+            # On older releases, we search for the most modern version
+            # available.
+            return _first_clang()
+    else:
+        return _first_clang([''] + _CLANG_SUFFIXES)
```

Comment:
Yup, gotcha. I'll rethink this based on this and your comments above. Instead of reusing a list of suffixes, I'll specify them separately for each function call. I'll also remove the default parameter from `_first_clang()`--using default array/dict parameters [opens up a lot of room for gotchas](http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments), so it's for the best.